### PR TITLE
Kindle/initialize session object + jest testing

### LINKED ISSUE #3: Initialize session object
This object is responsible for games that are actively being played.

Session objects last around an hour when there is no activity. When there is activity, the session max time will refresh itself, and the game can keep playing.

Sessions create their own threads once the host decides all the values are configured for game setup.
Each game lives in thread.

If the user tries to make a game while already in a thread, send an ephemeral to tell them to run the command in a channel

### PR DESCRIPTION
**Pull request by iSlammedMyKindle**
The session object takes care of the ongoing game. It's a straightforward object that simply holds hint count & a listener that is triggered when either the session expires, or `ended` is set to `true`.

It holds the following information:

* host
* guesser
* gameWin
* hints
* onEnd

onEnd is a custom listener that returns the session information back, allowing it to be passed to other functions, such as when the information to the database needs to be triggered (exp, general game information, etc)

# Jest

Jest has been initialized and configured to perform testing for any future files that may be created. We could use it to mock certain aspects of the code. We probably won't need to go into all modules, but testing core functionality will be useful.

### CODE REVIEW

File: tsconfig.json
Comment (iSlammedMyKindle) [inline tsconfig.json:1]: prettier caught wind of this file, but there are some actual changes made here. It's mainly so that things are importable for tests

File: tsconfig.json [hunk 1]
```diff
@@ -57,7 +57,7 @@
     // "outFile": "./",                                  /* Specify a file that bundles all outputs into one JavaScript file. If 'declaration' is true, also designates a file that bundles all .d.ts output. */
     "outDir": "./out",                                   /* Specify an output folder for all emitted files. */
     // "removeComments": true,                           /* Disable emitting comments. */
-    // "noEmit": true,                                   /* Disable emitting files from a compilation. */
+    "noEmit": true,                                      /* Disable emitting files from a compilation. */
```
Comment (MiguelHigueraDev) [inline tsconfig.json]: I don't understand this. As far as I am aware setting this to true makes it so when you compile TS it doesn't output anything.

Is there a reason for that?

File: test/session.test.ts
Comment (MiguelHigueraDev) [inline test/session.test.ts:1]: I think that the Arrange, Act, Assert comments could be removed because that's the pattern for all of them. Just a suggestion.

File: src/session.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
+/* Made by the cracklR Group in '24!
+Session objects help keep track of an ongoing game between two players.
+It expires after 1 hour, when it expires, the game will end on it's own, no EXP is awarded to either player.
+*/
+
+type onEndListener = (sessionObj: session) => void;
+
+export default class session {
```
Comment (MiguelHigueraDev) [inline src/session.ts]: The convention is that class names should be capitalized.

**Review by MiguelHigueraDev — CHANGES_REQUESTED** [review]
Thanks for the PR! Code looks pretty clean, I only have two suggestions and a question about a setting on `tsconfig.json`.

File: tsconfig.json (hunk 1 above)
Comment (iSlammedMyKindle) [reply tsconfig.json]: image
Yeah, so I had to toggle this in order to import with TS extensions.

perhaps we could have separate tsconfig files for testing & the src...

File: src/session.ts (hunk 2 above)
Comment (iSlammedMyKindle) [reply src/session.ts]: hmm... didn't realize that was a convention. If we want to do that we can

File: test/session.test.ts
Comment (iSlammedMyKindle) [reply test/session.test.ts:1]: This is a standard we use at our workplace. It's pretty consistent here, but arrange act & assert could get more sophisticated, so to have those markers would help to figure out where things are being setup versus where things are checked, etc.

Comment (MiguelHigueraDev) [reply test/session.test.ts:1]: Well we could keep it, it's not a huge deal.

File: tsconfig.json (hunk 1 above)
Comment (MiguelHigueraDev) [reply tsconfig.json]: Can't we just import it with .js extensions? That's what I do all the time and it works for me

File: src/session.ts (hunk 2 above)
Comment (MiguelHigueraDev) [reply src/session.ts]: Yeah I'd say it helps, so you know you are dealing with a class/interface/type/etc., instead of a regular variable (or class instance).

File: tsconfig.json (hunk 1 above)
Comment (iSlammedMyKindle) [reply tsconfig.json]: I'm trying to get it to work, but it seems to be struggling to properly infer things...
Let me try again, last time I went for it was last night

File: jest.config.js
Comment (iSlammedMyKindle) [inline jest.config.js:1]: Explanation about this file @MiguelHigueraDev :

This is a file specific for unit testing. It basically acts like a tsconfig specific to jest.

You were correct in that the `.js` extension works just fine with imports when running `npx tsc`, but for whatever reason I could not get that working with jest. This is is mainly the workaround since from what I understand, jest will compile & interpret typescript in real time. It's different from the base imports, but this case is peculiar because we aren't using babel to compile this into JS, rather it's just native typescript compiling into EsNext

Comment (MiguelHigueraDev) [reply jest.config.js:1]: Ahh alright, thanks for the explanation and for fixing it!

**Review by MiguelHigueraDev — APPROVED** [review]
